File: PythonCode/crowdsignals_ch2.py
##############################################################
#                                                            #
#    Mark Hoogendoorn and Burkhardt Funk (2017)              #
#    Machine Learning for the Quantified Self                #
#    Springer                                                #
#    Chapter 2                                               #
#                                                            #
##############################################################


dataset_path = '../datasets/crowdsignals.io/csv-participant-one/'
result_dataset_path = './intermediate_datafiles/'

# Import the relevant classes.

from Chapter2.CreateDataset import CreateDataset
from util.VisualizeDataset import VisualizeDataset
from util import util
import copy
import os


if not os.path.exists(result_dataset_path):
    print('Creating result directory: ' + result_dataset_path)
    os.makedirs(result_dataset_path)

# Chapter 2: Initial exploration of the dataset.

# Set a granularity (i.e. how big are our discrete time steps). We start very
# coarse grained, namely one measurement per minute, and secondly use four measurements
# per second

granularities = [60000, 250]
datasets = []

for milliseconds_per_instance in granularities:

    # Create an initial dataset object with the base directory for our data and a granularity
    DataSet = CreateDataset(dataset_path, milliseconds_per_instance)

    # Add the selected measurements to it.

    # We add the accelerometer data (continuous numerical measurements) of the phone and the smartwatch
    # and aggregate the values per timestep by averaging the values/
    DataSet.add_numerical_dataset('accelerometer_phone.csv', 'timestamps', ['x','y','z'], 'avg', 'acc_phone_')
    DataSet.add_numerical_dataset('accelerometer_smartwatch.csv', 'timestamps', ['x','y','z'], 'avg', 'acc_watch_')

    # We add the gyroscope data (continuous numerical measurements) of the phone and the smartwatch
    # and aggregate the values per timestep by averaging the values/
    DataSet.add_numerical_dataset('gyroscope_phone.csv', 'timestamps', ['x','y','z'], 'avg', 'gyr_phone_')
    DataSet.add_numerical_dataset('gyroscope_smartwatch.csv', 'timestamps', ['x','y','z'], 'avg', 'gyr_watch_')

    # We add the heart rate (continuous numerical measurements) and aggregate by averaging again
    DataSet.add_numerical_dataset('heart_rate_smartwatch.csv', 'timestamps', ['rate'], 'avg', 'hr_watch_')

    # We add the labels provided by the users. These are categorical events that might overlap. We add them
    # as binary attributes (i.e. add a one to the attribute representing the specific value for the label if it
    # occurs within an interval).
    DataSet.add_event_dataset('labels.csv', 'label_start', 'label_end', 'label', 'binary')

    # We add the amount of light sensed by the phone (continuous numerical measurements) and aggregate by averaging again
    DataSet.add_numerical_dataset('light_phone.csv', 'timestamps', ['lux'], 'avg', 'light_phone_')

    # We add the magnetometer data (continuous numerical measurements) of the phone and the smartwatch
    # and aggregate the values per timestep by averaging the values
    DataSet.add_numerical_dataset('magnetometer_phone.csv', 'timestamps', ['x','y','z'], 'avg', 'mag_phone_')
    DataSet.add_numerical_dataset('magnetometer_smartwatch.csv', 'timestamps', ['x','y','z'], 'avg', 'mag_watch_')

    # We add the pressure sensed by the phone (continuous numerical measurements) and aggregate by averaging again
    DataSet.add_numerical_dataset('pressure_phone.csv', 'timestamps', ['pressure'], 'avg', 'press_phone_')

    # Get the resulting pandas data table

    dataset = DataSet.data_table

    # Plot the data

    DataViz = VisualizeDataset()

    # Boxplot
    DataViz.plot_dataset_boxplot(dataset, ['acc_phone_x','acc_phone_y','acc_phone_z','acc_watch_x','acc_watch_y','acc_watch_z'])

    # Plot all data
    DataViz.plot_dataset(dataset, ['acc_', 'gyr_', 'hr_watch_rate', 'light_phone_lux', 'mag_', 'press_phone_', 'label'], ['like', 'like', 'like', 'like', 'like', 'like', 'like','like'], ['line', 'line', 'line', 'line', 'line', 'line', 'points', 'points'])

    # And print a summary of the dataset

    util.print_statistics(dataset)
    datasets.append(copy.deepcopy(dataset))

# And print the table that has been included in the book

util.print_latex_table_statistics_two_datasets(datasets[0], datasets[1])

# Finally, store the last dataset we have generated (250 ms).
dataset.to_csv(result_dataset_path + 'chapter2_result.csv')

File: RCode/crowdsignals_ch2.R
##########################################################################
#
#   Mark Hoogendoorn & Burkhardt Funk (2017)
#   Machine Learning for the Quantified Self
#   Springer
#   Chapter 2 - Sensory Data
#   ./crowdsignals_ch2.R
# 
##########################################################################

# Import libraries and source utils
rm(list=ls())
require(data.table)
require(bit64)
require(plotly)
require(reshape2)
source("./chapter2/createDataset.R")
source("./chapter2/visiualizeData.R")

# Set a granularity (i.e. how big are our discrete time steps). We start very
# coarse grained, namely one measurement per minute, and secondly use four measurements
# per second
datasetPath = "../datasets/crowdsignals.io/csv-participant-one/"
#granularities = c(60000,250)
granularities = c(250)

boxPlot = NULL 
linePlot = NULL
for (timeWindow in granularities) {
  
  # We add the accelerometer data (continuous numerical measurements) of the phone and the smartwatch
  # and aggregate the values per timestep by averaging the values/
  crowdSignalData = addNumericalDataset(NULL, "accelerometer_phone.csv","timestamps", c("x","y","z"), "avg", "acc_phone_", datasetPath,timeWindow)
  crowdSignalData = addNumericalDataset(crowdSignalData, "accelerometer_smartwatch.csv","timestamps", c("x","y","z"), "avg", "acc_watch_", datasetPath,timeWindow)
  
  # We add the gyroscope data (continuous numerical measurements) of the phone and the smartwatch
  # and aggregate the values per timestep by averaging the values
  crowdSignalData = addNumericalDataset(crowdSignalData, "gyroscope_phone.csv","timestamps", c("x","y","z"), "avg", "gyr_phone_", datasetPath,timeWindow)
  crowdSignalData = addNumericalDataset(crowdSignalData, "gyroscope_smartwatch.csv","timestamps", c("x","y","z"), "avg", "gyr_watch_", datasetPath,timeWindow)
  
  # We add the heart rate (continuous numerical measurements) and aggregate by averaging again
  crowdSignalData = addNumericalDataset(crowdSignalData, "heart_rate_smartwatch.csv","timestamps", c("rate"), "avg", "hr_watch_", datasetPath,timeWindow)
  
  # We add the labels provided by the users. These are categorical events that might overlap. We add them
  # as binary attributes (i.e. ad a one to the attribute representing the specific value for the label if it
  # occurs within an interval).
  crowdSignalData = addEventDataset(crowdSignalData, "labels.csv", "label_start", "label_end", "label", "binary", datasetPath)
  
  # We add the amount of light sensed by the phone (continuous numerical measurements) and aggregate by averaging again
  crowdSignalData = addNumericalDataset(crowdSignalData, "light_phone.csv","timestamps", c("lux"), "avg", "light_phone_", datasetPath,timeWindow)
  
  # We add the magnetometer data (continuous numerical measurements) of the phone and the smartwatch
  # and aggregate the values per timestep by averaging the values/
  crowdSignalData = addNumericalDataset(crowdSignalData, "magnetometer_phone.csv","timestamps", c("x","y","z"), "avg", "mag_phone_", datasetPath,timeWindow)
  crowdSignalData = addNumericalDataset(crowdSignalData, "magnetometer_smartwatch.csv","timestamps", c("x","y","z"), "avg", "mag_watch_", datasetPath,timeWindow)
  
  # We add the pressure sensed by the phone (continuous numerical measurements) and aggregate by averaging again
  crowdSignalData = addNumericalDataset(crowdSignalData, "pressure_phone.csv","timestamps", c("pressure"), "avg", "press_phone_", datasetPath,timeWindow)
  crowdSignalData$time = as.POSIXct(crowdSignalData$time/1000, origin="1970-01-01")
  
  # Lineplot
  # to plot, simply use print, e.g. "print(p[1])"
  temp = plotDataset(crowdSignalData, c("acc_", "gyr_", "hr_watch_rate", "light_phone_lux", "mag_", "press_phone_","label"),NULL)
  if (is.null(linePlot)) linePlot = subplot(temp,nrows = length(temp),shareX = TRUE) else
    linePlot = list(linePlot,subplot(temp,nrows = length(temp),shareX = TRUE))
  
  # Boxplot
  if(is.null(boxPlot)) boxPlot = list(plotBoxplot(crowdSignalData, c("acc_phone_x","acc_phone_y","acc_phone_z","acc_watch_x","acc_watch_y","acc_watch_z"))) else
    boxPlot = list(boxPlot,plotBoxplot(crowdSignalData, c("acc_phone_x","acc_phone_y","acc_phone_z","acc_watch_x","acc_watch_y","acc_watch_z")))   
}

# Only the dataset with the last value for granularities is saved 
resultPath = "./intermediate_datafiles/"
save(crowdSignalData, file=paste(resultPath,"chapter2_result.RData",sep=""))

# Save as CSV file

# crowdSignalData$time = as.character(crowdSignalData$time,format="%Y-%m-%d %H:%M:%OS4")
# write.csv(x=crowdSignalData, paste(resultPath,"chapter2_result.csv",sep=""))